cmd/av/stack_sync.go

```diff
@@ -152,64 +256,61 @@ base branch.
 		}
 
 		_, _ = fmt.Fprintf(os.Stderr, "Stack sync complete\n")
+
+		// Return to the starting branch when we're done
+		if _, err := repo.CheckoutBranch(&git.CheckoutBranch{
+			Name: state.OriginalBranch,
+		}); err != nil {
+			return err
+		}
+
 		return nil
 	},
 }
 
-// doStackSync performs the stack sync operation.
-// It returns a boolean indicating whether or not the sync was interrupted
-// (if true, the stack sync can be resumed with the --continue flag).
-// TODO[mvp]:
-//    This whole function is kind of a hot mess, and a lot of it is rooted in
-//    how awkward the API is for dealing with "branch trees" (terminology is
-//    hard when you're creating a meta-tree-structure on top of git branches).
-//    Let's refactor that a bit and then re-work this.
-func doStackSync(repo *git.Repo, config stackSyncConfig) (bool, error) {
-	branches, err := meta.ReadAllBranches(repo)
-	if err != nil {
-		return false, err
-	}
+type syncResult struct {
+	Conflict      bool
+	CurrentBranch string
+}
 
-	currentBranch, err := repo.CurrentBranchName()
+// Find all the ancestor branches of the given branch name and append them to
+// the given slice (in topological order: a comes before b if a is an ancestor
+// of b).
+func previousBranches(branches map[string]meta.Branch, name string) ([]string, error) {
+	current, ok := branches[name]
+	if !ok {
+		return nil, errors.Errorf("branch metadata not found for %q", name)
+	}
+	if current.Parent == "" {
+		return nil, nil
+	}
+	previous, err := previousBranches(branches, current.Parent)
 	if err != nil {
-		return false, err
+		return nil, err
 	}
+	return append(previous, current.Parent), nil
```

Comment:
does this take care of nil returned back from the recursive call?

Reply:
Yeah. Go is weird with slices, but appending to a nil slice initializes it as a slice with one element.